### ISSUE TITLE
Serenity/JS should vend @serenity-js/serenity-bdd package with serenity-bdd jar included

### ISSUE THREAD

**Issue by fapspirit** [serenity-js/serenity-js#2560]
### What's the problem you're trying to solve?

When building project in environments with restricted network access, it's hard to setup serenity-bdd reporter as it requires access to maven repo. There are workarounds available, but since version of serenity-bdd jar file is fixed in the code constants, I think it's worth to include it into the package artifact that is available in npm registry.

I believe this might also simplify overall setup experience, while keeping current customization capabilities.

### How would you like to solve it?

This might be solved in a convenient way similar to `allure-commandline` package: https://github.com/allure-framework/allure-npm/blob/main/.github/workflows/npm-publish.yml

Download jar file as part of the build step of the `@serenity-js/serenity-bdd` package and include it into build artifact when releasing to npm.

### Are there any alternatives?

These are workarounds that I came up to:
- Commit serenity-bdd jar file into git repo of a project. Downside is that serenity-bdd jar file is ~60mb so this will influence the size of the git repo
- Create a separate package similar to `allure-commandline`. Downside is synchronization of the versions and additional frictions
- Setup a proxy maven repo available within restricted network (similar to npm registry proxy like Verdaccio). Downside is additional frictions required

### How can we make it happen?

- [X] [work on this myself](https://serenity-js.org/contributing/#propose-an-enhancement) (with Serenity/JS team guidance)
- [ ] [hire the Serenity/JS team](https://github.com/sponsors/serenity-js) to build it for me
- [X] add it to the Serenity/JS backlog and wait for a volunteer

**Comment by jan-molak** [comment]
Hey @fapspirit, so that's an interesting problem. But let's start with the current solution.

## The current solution

Network-restricted environments typically have access to a local artifact registry like Artifactory, Nexus, or similar.

You can instruct the `@serenity-js/serenity-bdd` downloader to download the jar from a local registry by providing a `--repository` parameter, as per the [documentation](https://serenity-js.org/handbook/reporting/serenity-bdd-reporter/#downloading-serenity-bdd-cli):

```
npx serenity-bdd update --repository https://artifactory.example.org/
```

You can also instruct the downloader to download a specific version of the Serenity BDD artifact if you want to use a newer version than what Serenity/JS supports, or if you have a local build:

```
npx serenity-bdd update --repository https://artifactory.example.org/ --artifact net.serenity-bdd:serenity-cli:jar:4.1.20
```

Since Serenity BDD tends to release less frequently than Serenity/JS, this separation allows you download less data and retain the cached jar between builds. The price for that is a slightly more involved setup that typically requires [adding a `postinstall` or `pretest` script](https://serenity-js.org/handbook/reporting/serenity-bdd-reporter/#using-serenity-bdd-cli) to your `package.json`.

Would this work for you?

## Alternative 1 - embed the Serenity BDD jar in `@serenity-js/serenity-bdd`

As you suggest, we could embed the Serenity BDD jar in the `@serenity-js/serenity-bdd` node module. It will definitely make the setup easier, but has the following limitations:
- you could no longer change the version of the Serenity BDD jar independently of the node module
- you could no longer cache the jar locally between installations
- since Serenity BDD releases on a different and less frequent schedule, multiple builds of `@serenity-js/serenity-bdd` node module would contain the same version of the Serenity BDD jar

## Alternative 2 - introduce a wrapper module for Serenity BDD jar

To avoid the above limitations, we could introduce a new module like `@serenity-js/serenity-bdd-cli` that:
- is separate from `@serenity-js/serenity-bdd`
- embeds the Serenity BDD jar
- does not conform to the [Serenity/JS versioning strategy](https://serenity-js.org/releases/versioning/), but instead reflects the version of the embedded jar so that, for example, Serenity/JS 3.29.2 could depend on `@serenity-js/serenity-bdd-cli` 4.1.20 

The downside of this approach would be that developers:
- would need to add a new module to their `package.json`
- the version numbers might be confusing

```json
{
  "devDependencies": {
    "@serenity-js/core": "3.29.2",
    "@serenity-js/assertions": "3.29.2",
    "@serenity-js/rest": "3.29.2",
    "@serenity-js/playwright": "3.29.2",
    "@serenity-js/playwright-test": "3.29.2",
    "@serenity-js/serenity-bdd": "3.29.2",
    "@serenity-js/serenity-bdd-cli": "4.1.20",       // <-- seems like an odd one out
    "@serenity-js/web": "3.29.2"
  }
}
```

## Alternative 3 - write a new HTML reporter

Another alternative would be to write a completely new HTML reporter for Serenity/JS, but without a dependency on Java, and then potentially reuse it for Serenity BDD.
This seems like quite a bit of work, though...

---


Thoughts?

**Comment by fapspirit** [comment]
Hey @jan-molak, thank you for a detailed answer! Here are my thoughts regarding it:

> Network-restricted environments typically have access to a local artifact registry like Artifactory, Nexus, or similar.

Typically this is correct, although it's not always the case. I.e. when using verdaccio, which supports only npm registry. Or even worse, some weird custom solution. In my experience I might say that more weird a proxy setup is, the more likely it's hard to setup an additional dedicated artifact registry. This is probably not a blocker though as for the weirdest setup this can be eliminated by pushing a Serenity BDD jar file straight to the git repo (which is what I did for my situation).

> Since Serenity BDD tends to release less frequently than Serenity/JS, this separation allows you download less data and retain the cached jar between builds. The price for that is a slightly more involved setup that typically requires [adding a postinstall or pretest script](https://serenity-js.org/handbook/reporting/serenity-bdd-reporter/#using-serenity-bdd-cli) to your package.json.

If I understand correctly, these two sentences actually contradicts each other. If using `postinstall` script, the Serenity BDD jar file will be downloaded on each install anyway (unless another location for cache is specified), so it's pretty much the same effect as I initially proposed. Going further, with current package managers like pnpm, it's more likely to download less data than with the current setup due to package cache on a machine (i.e. [pnpm storage](https://github.com/pnpm/pnpm?tab=readme-ov-file#background)), as current approach involves http request to another storage on each clean install.

> you could no longer change the version of the Serenity BDD jar independently of the node module

I don't think implementation of built-in Serenity BDD is a replacement for a current approach which supports customizations. More like the opposite, it's a progressive enhancement with a new default that just eliminates this additional step of downloading and managing Serenity BDD jar file for a default setup. But if one would like to override it, I don't see a reason not to allow to do so with the current logic.

> you could no longer cache the jar locally between installations

As I mentioned previously, it's all depends on the cache of node_modules, but also we shouldn't remove current ability to explicitly set a path to a directory with cache or version of the Serenity BDD, so it won't break anything.

> since Serenity BDD releases on a different and less frequent schedule, multiple builds of @serenity-js/serenity-bdd node module would contain the same version of the Serenity BDD jar

Could you please elaborate on what is the problem here? The only use case I see is that if a new version of `@serenity-js/serenity-bdd` is released, once it's updated in the project, we won't download a new version of Serenity BDD jar file, but in fact, if the default cache dir is used (which is a dir inside @serenity-js/serenity-bdd package in node_modules), it will override it anyway. 

> To avoid the above limitations, we could introduce a new module like @serenity-js/serenity-bdd-cli
> [...]
> The downside of this approach would be that developers:
> - would need to add a new module to their package.json
> - the version numbers might be confusing

I think such package is just a slightly better than a current approach, and I think you're right about the confusion. If I understand correctly, it means separating `update` command from `@serenity-js/serenity-bdd` package into a new package and some of these package should be aware of another package to wire-up the paths to a jar file, but I don't quite understand the reason why it should be done? My main concern here is that `@serenity-js/serenity-bdd` does not make much sense without Serenity BDD jar file, so it's does make sense to vend them together.

> Alternative 3 - write a new HTML reporter
> Another alternative would be to write a completely new HTML reporter for Serenity/JS, but without a dependency on Java, and then potentially reuse it for Serenity BDD.
This seems like quite a bit of work, though...

This is indeed a lot of work and I don't think this makes sense unless there are clear benefits of a custom implementation. Although it'll eliminate the need to manage Java env and all the related stuff, let's omit this idea for now.

**Comment by jan-molak** [comment]
> > Network-restricted environments typically have access to a local artifact registry like Artifactory, Nexus, or similar.

> Typically this is correct, although it's not always the case. I.e. when using verdaccio, which supports only npm registry.

Interesting! It's a good argument. I think I've only seen Verdaccio used with Nx workspaces so far. Have you seen it being used outside of this context?

> If using postinstall script, the Serenity BDD jar file will be downloaded on each install anyway (unless another location for cache is specified), so it's pretty much the same effect as I initially proposed.

I think I'm starting to see your point; So what you're saying is that if `@serenity-js/serenity-bdd` shipped with the Serenity BDD jar embedded in it, you could cache the whole thing locally. Even without Verdaccio, you could potentially use the "standard" npm to create local cache and instruct your CI to maintain it between pipeline runs. For example:

```
npm ci --cache .npm --prefer-offline
```

And it's true that `pnpm` could make that even easier... Hmm 🤔 

> But if one would like to override it, I don't see a reason not to allow to do so with the current logic.

Please walk me through it. Let's say `@serenity-js/serenity-bdd` ships with the jar embedded in the node module, which means the jar is published to `npmjs.com`, propagated to local company mirrors like Artifactory, etc.
How would people override what jar they want to use? Are you thinking of keeping the downloader functionality, or are you proposing to remove that altogether, but keep the `--artifact` option for the `serenity-bdd run` command?

> As I mentioned previously, it's all depends on the cache of node_modules, but also we shouldn't remove current ability to explicitly set a path to a directory with cache or version of the Serenity BDD, so it won't break anything.

Do you have any mechanism in mind to make both embedding and downloading work simultaneously?

**Comment by fapspirit** [comment]
> > This is probably not a blocker though as for the weirdest setup this can be eliminated by pushing a Serenity BDD jar file straight to the git repo (which is what I did for my situation).

> I agree, committing a jar to git is suboptimal. I'd love to learn more about the setup that required it if you're OK with sharing?

Unfortunately, I can't disclosure any specific details, but the problem is that we don't have any registries like Artifactory or Nexus and we can't have them due to some security considerations.

> > > Network-restricted environments typically have access to a local artifact registry like Artifactory, Nexus, or similar.

> > Typically this is correct, although it's not always the case. I.e. when using verdaccio, which supports only npm registry.

> Interesting! It's a good argument. I think I've only seen Verdaccio used with Nx workspaces so far. Have you seen it being used outside of this context?

I'm not aware of any open-source projects, but In some of previous companies that I worked for we had a Verdaccio registry deployed as a proxy for all 3-rd party packages so we won't suffer due to outages of npm or any other network-related problems. In one company we hosted our 1-st party packages in the self-hosted Gitlab, but in another company we had a custom source version control system so we used Verdaccio to publish our 1-st party packages as well.

I believe nowadays there are plenty of proxy registries to spin up in the internal network, but I'm not sure wether it's worth to setup just for a single package?

> > If using postinstall script, the Serenity BDD jar file will be downloaded on each install anyway (unless another location for cache is specified), so it's pretty much the same effect as I initially proposed.

> I think I'm starting to see your point; So what you're saying is that if @serenity-js/serenity-bdd shipped with the Serenity BDD jar embedded in it, you could cache the whole thing locally. Even without Verdaccio, you could potentially use the "standard" npm to create local cache and instruct your CI to maintain it between pipeline runs. For example:
```
npm ci --cache .npm --prefer-offline
```
> And it's true that pnpm could make that even easier... Hmm 🤔

Yes, this is a benefit that basically comes for free when using caching of node_modules in some way. Current package managers are really good at caching, so why not to utilize their capabilities?

> > But if one would like to override it, I don't see a reason not to allow to do so with the current logic.

> Please walk me through it. Let's say @serenity-js/serenity-bdd ships with the jar embedded in the node module, which means the jar is published to npmjs.com, propagated to local company mirrors like Artifactory, etc.
How would people override what jar they want to use? Are you thinking of keeping the downloader functionality, or are you proposing to remove that altogether, but keep the --artifact option for the serenity-bdd run command?

Sure, let me elaborate on the idea. First of all, we don't need to change anything in the current workflow so we can leave current APIs as they are: leave `update` command as is and `run` command as is as well, with all of the flags that they're support (e.g. `--source`, `--destination`, `--cacheDir`, etc.). Current flow looks like this: once I installed serenity-js packages, I need to run `serenity-bdd update` to populate cache dir inside `node_modules/@serenity-js/serenity-bdd/cache` with the `serenity-cli-4.1.20.jar` file. So the idea is to basically get rid of the second step for a default setup: why `@serenity-js/serenity-bdd` can't be installed with `cache/serenity-cli-4.1.20.jar` file already included? Since nothing has been changed in the logic of the `@serenity-js/serenity-bdd`, it's still possible to leverage the same API to download a specific version of Serenity BDD jar file or to set a specific cache directory.

This idea made me think of another question: how many users actually needs to override version of Serenity BDD file? I don't have any data, but my guess is that majority of users needs just a default configuration. Please feel free to correct me, as I might be biased.

> > As I mentioned previously, it's all depends on the cache of node_modules, but also we shouldn't remove current ability to explicitly set a path to a directory with cache or version of the Serenity BDD, so it won't break anything.

> Do you have any mechanism in mind to make both embedding and downloading work simultaneously?

I've tried to explain the idea in my previous comment. Feel free to ask any additional questions!

> > My main concern here is that @serenity-js/serenity-bdd does not make much sense without Serenity BDD jar file, so it's does make sense to vend them together.

> There's one more use case we haven't discussed. This is a JVM project using Serenity/JS for the front end, Serenity BDD for the back end, and the Serenity BDD Maven plugin to produce an aggregated report. In this scenario, Serenity BDD CLI is not used at all, and the role of the @serenity-js/serenity-bdd module is reduced to producing .json reports, to be ingested by the Maven plugin.

Thank you for clarifications! My point is these .json files are useless without the Maven plugin since as a user I would like to see a html report, which is generated from these .json files. But I'm not really familiar with the Java ecosystem, so now I see a use case when there are both backend API written in Java and frontend UI written in JS/TS in a single project, in that case being able to set a cache dir for Serenity BDD Maven plugin is indeed a good way to go. My experience with Serenity BDD is scoped to the frontend-only projects.

**Comment by jan-molak** [comment]
I was thinking about your proposal some more. You're right that it could be implemented as a progressive enhancement to the current setup, so let's discuss how we can make it happen.

The main thing that would need to change is that the GitHub Actions pipeline would need to call `serenity-bdd update` _at some point_ before publishing the artifacts to NPM. Assuming the `.jar` is in the `cache` directory, it should get automatically included in the published artifact since it's not explicitly excluded by [.npmignore](https://github.com/serenity-js/serenity-js/blob/d1bffbbf6e4469eeec24f86784a5d60b90202910/packages/serenity-bdd/.npmignore).

The question is when and how to call `serenity-bdd update` in the [pipeline](https://github.com/serenity-js/serenity-js/blob/d1bffbbf6e4469eeec24f86784a5d60b90202910/.github/workflows/main.yaml).

## Compile stage

The earliest point in the pipeline where we could call `serenity-bdd update` is the [`compile` stage](https://github.com/serenity-js/serenity-js/blob/d1bffbbf6e4469eeec24f86784a5d60b90202910/.github/workflows/main.yaml#L29)

This could be done either by:
- introducing a `postinstall` script in [`@serenity-js/serenity-bdd`](https://github.com/serenity-js/serenity-js/blob/d1bffbbf6e4469eeec24f86784a5d60b90202910/packages/serenity-bdd/package.json#L37), calling `serenity-bdd update`
- modifying the `compile` script to call `serenity-bdd update`

It feels like adding a `postinstall` could be cleaner; I have avoided doing that in the past since NPM calls `postinstall` both at the build time and during consumer-side installation, and if the consumer uses an internal artifact registry, calling the `postinstall` script with default settings would fail. However, if `@serenity-js/serenity-bdd` shipped with the `.jar`, `postinstall` would not make any external calls since the `.jar` would have already been present.

If we added a `postinstall` script that downloaded the `.jar`, we'd also need to make sure to include the `.jar` in the [package of artifacts](https://github.com/serenity-js/serenity-js/blob/d1bffbbf6e4469eeec24f86784a5d60b90202910/.github/workflows/main.yaml#L31) propagated through the pipeline

## Integration test stage

[Serenity/JS integration tests use Serenity BDD](https://github.com/search?q=repo%3Aserenity-js%2Fserenity-js+%22serenity-bdd+update%22+path%3A%2F%5Eintegration%5C%2F%2F&type=code), so we could either:
- leave the explicit call to `serenity-bdd update` in their respective `package.json` files
- remove it and rely on the pre-cached .jar
- or introduce a separate integration test module to verify explicitly calling the downloader

The last option might be an overkill, though, as we'd call the downloader during the `compile` step in the pipeline anyway.

**Comment by fapspirit** [comment]
Your plan makes sense to me, let's give it a try. Since it's not a breaking change, if installing via `postinstall` script won't be ideal, it's still possible to change it in the future to an explicit step in the pipeline.

For integration test stage, I'd personally prefer the second option as it's in fact the new default behaviour. And it's actually implicitly checks correctness of `serenity-bdd update` command that is executed as part of the `compile` stage: the `serenity-bdd run` command will fail if there are no jar file, which is supposed to be added as part of the `compile` stage.

As for potential challenges, I don't think the size of a package is an issue for npmjs.org. Here is the biggest package in the registry: https://www.npmjs.com/package/@chainsafe/eth2-spec-tests (size is 5.6 GB!).

The amount of changes to be done seems pretty small. Do you want me to work on them or you plan to implement them yourself?

**Comment by jan-molak** [comment]
> Your plan makes sense to me, let's give it a try. Since it's not a breaking change, if installing via postinstall script won't be ideal, it's still possible to change it in the future to an explicit step in the pipeline.

Yes, exactly what I was thinking.

> For integration test stage, I'd personally prefer the second option as it's in fact the new default behaviour. And it's actually implicitly checks correctness of serenity-bdd update command that is executed as part of the compile stage: the serenity-bdd run command will fail if there are no jar file, which is supposed to be added as part of the compile stage.

Sounds good to me

> As for potential challenges, I don't think the size of a package is an issue for npmjs.org. Here is the biggest package in the registry: https://www.npmjs.com/package/@chainsafe/eth2-spec-tests (size is 5.6 GB!).

Wow, just... wow.

> Do you want me to work on them 

Yes, please, thank you! Any issue, just let me know.

If you could also use [`npm run commit`](https://serenity-js.org/community/contributing/commit-message-conventions/) to indicate this is a new `feature`, the pipeline will bump the minor version number on release.

**Comment by jan-molak** [comment]
#2591 was released with Serenity/JS 3.30.0

Outstanding documentation updates (so that I don't forget):
- https://serenity-js.org/handbook/reporting/serenity-bdd-reporter/
- remove Serenity BDD-related proxy setup instructions from all the project template readme files
